test/test_type_promotion.py

```diff
@@ -343,19 +343,69 @@ def test_create_bool_tensors(self, device):
         # this seems like odd behavior but ints also create float tensors, numpy doesn't have this function.
         self.assertEqual(torch.scalar_tensor(False, device=device), torch.tensor(0., device=device))
 
+    @dtypes(*itertools.product(torch.testing.get_all_dtypes(), torch.testing.get_all_dtypes()))
+    def test_result_type(self, device, dtypes):
+        "Test result_type for tensor vs tensor and scalar vs scalar."
+
+        def _get_dtype(x):
+            "Get the dtype of x if x is a tensor. If x is a scalar, get its corresponding dtype if it were a tensor."
+            if torch.is_tensor(x):
+                return x.dtype
+            elif isinstance(x, bool):
+                return torch.bool
+            elif isinstance(x, int):
+                return torch.int64
+            elif isinstance(x, float):
+                return torch.float32
+            elif isinstance(x, complex):
+                return torch.complex64
+            else:
+                raise AssertionError(f"Unkonwn type {x}")
+
+        # tensor against tensor
+        a_tensor = torch.tensor((0, 1), device=device, dtype=dtypes[0])
+        a_single_tensor = torch.tensor(1, device=device, dtype=dtypes[0])
+        a_scalar = a_single_tensor.item()
+        b_tensor = torch.tensor((1, 0), device=device, dtype=dtypes[1])
+        b_single_tensor = torch.tensor(1, device=device, dtype=dtypes[1])
+        b_scalar = b_single_tensor.item()
+        for a, b in ((a_tensor, b_tensor), (a_single_tensor, b_single_tensor), (a_scalar, b_scalar)):
```

Comment:
This is only testing tensor x tensor, zero dim tensor x zero dim tensor, scalar x scalar, but we want to test the cross of these, like tensor x scalar.

Reply:
tensor x scalar is in the spot checking tests, because it doesn't follow the usual rule followed by `promote_types`.

Reply:
That's right. See comment below about skipping `promote_types` in those cases. `result_type` should work on them, however.